upgrade/v110to111/upgrade.go

```diff
@@ -82,3 +148,409 @@ func doLonghornNodeUpgrade(namespace string, lhClient *lhclientset.Clientset) (e
+	}
+	return nil
+}
+
+func upgradeEngineImages(namespace string, lhClient *lhclientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"upgrade engine image failed")
+	}()
+
+	eiList, err := lhClient.LonghornV1beta1().EngineImages(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, ei := range eiList.Items {
+		if err := upgradeLabelsForEngineImage(&ei, lhClient, namespace); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func upgradeLabelsForEngineImage(ei *longhorn.EngineImage, lhClient *lhclientset.Clientset, namespace string) (err error) {
+	metadata, err := meta.Accessor(ei)
+	if err != nil {
+		return err
+	}
+	engineImageLabels := metadata.GetLabels()
+	newEngineImageLabels := upgradeutil.MergeStringMaps(engineImageLabels, types.GetEngineImageLabels(ei.Name))
+	if reflect.DeepEqual(engineImageLabels, newEngineImageLabels) {
+		return nil
+	}
+	metadata.SetLabels(newEngineImageLabels)
+	if _, err := lhClient.LonghornV1beta1().EngineImages(namespace).Update(ei); err != nil {
+		return err
+	}
+	return nil
+}
+
+func upgradeShareManagerPods(namespace string, lhClient *lhclientset.Clientset, kubeClient *clientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"upgrade share manager pods failed")
+	}()
+	smPods, err := upgradeutil.ListShareManagerPods(namespace, kubeClient)
+	if err != nil {
+		return err
+	}
+	for _, pod := range smPods {
+		sm, err := lhClient.LonghornV1beta1().ShareManagers(namespace).Get(types.GetShareManagerNameFromShareManagerPodName(pod.Name), metav1.GetOptions{})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				continue
+			}
+			return err
+		}
+		if err := upgradeShareManagerPodsLabels(&pod, sm, kubeClient, namespace); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func upgradeShareManagerPodsLabels(pod *v1.Pod, sm *longhorn.ShareManager, kubeClient *clientset.Clientset, namespace string) (err error) {
+	metadata, err := meta.Accessor(pod)
+	if err != nil {
+		return err
+	}
+	podLabels := metadata.GetLabels()
+	newPodLabels := upgradeutil.MergeStringMaps(podLabels, types.GetShareManagerLabels(sm.Name, sm.Spec.Image))
+	if reflect.DeepEqual(podLabels, newPodLabels) {
+		return nil
+	}
+	metadata.SetLabels(newPodLabels)
+	if _, err := kubeClient.CoreV1().Pods(namespace).Update(pod); err != nil {
+		return err
+	}
+	return nil
+}
+
+func upgradeInstanceManagerPods(namespace string, kubeClient *clientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"upgrade instance manager pods failed")
+	}()
+	imPods, err := upgradeutil.ListIMPods(namespace, kubeClient)
+	if err != nil {
+		return err
+	}
+	for _, pod := range imPods {
+		if err := updateIMPodLastAppliedTolerationsAnnotation(&pod, kubeClient, namespace); err != nil {
+			return err
+		}
+		// Note that we already upgrade instance manager pods' labels in v100to101, so no need to upgrade it here
```

Comment:
Changing the functionality of old upgrade code is kinda hmm, but since we didn't make the update implementations immutable (i.e. freeze the code for updates). This is okay since it saves some extra update calls.